designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,527 @@
+
+### Loss of standarized way to expose configs
+
+First, it's difficult to know how a plugin exports configurations. A configuration can be an object or an array, and it may be accessible via a `configs` key or via a module specifier. Further, typescript-eslint actually [recommends](https://typescript-eslint.io/packages/typescript-eslint/#flat-config-extends) the use of their own utility to create configs, which [confuses users](https://github.com/typescript-eslint/typescript-eslint/issues/8496). That means you can have multiple ways of importing a config into your own. Here's an example:
+
+```js
+import js from "@eslint/js";
+import tailwind from "eslint-plugin-tailwindcss";
+import reactPlugin from "eslint-plugin-react";
+import eslintPluginImportX from 'eslint-plugin-import-x'
+
+export default [
+    js.configs.recommended,
+    ...tailwind.configs["flat/recommended"],
+    ...reactPlugin.configs.flat.recommended,
+    eslintPluginImportX.flatConfigs.recommended,
+];
+```
+
+### Confusion over `...` syntax
+
+As discussed in the previous section, configs may be objects or arrays, and the latter requires the use of the spread syntax (`...`) to include its contents into the overall array. We've found that JavaScript beginners don't know or understand spread syntax, making this distinction between objects and arrays even more confusing. 
+
+### Difficult extending configs for specific file patterns
+
+Further adding to the confusion over objects and arrays is how to take an existing config and apply it to a subset of files. This explicitly requires you to know if something is an object or array because how you extend it varies depending on the type. For objects, you do this:
+
+```js
+// eslint.config.js
+import js from "@eslint/js";
+
+export default [
+    {
+        ...js.configs.recommended,
+        files: ["**/src/safe/*.js"]
+    }
+];
+```
+
+For arrays you do this:
+
+```js
+// eslint.config.js
+import exampleConfigs from "eslint-config-example";
+
+export default [
+    ...exampleConfigs.map(config => ({
+        ...config,
+        files: ["**/src/safe/*.js"]
+    })),
+
+    // your modifications
+    {
+        rules: {
+            "no-unused-vars": "warn"
+        }
+    }
+];
+```
+
+In both cases, users need to use the spread syntax, and in both cases, just glancing at the config object doesn't really tell you what exactly is happening. It's a lot of syntax for what used to be a single statement in eslintrc.
+
+## Goals
+
+There are two goals with this design:
+
+1. **Make it easier to configure ESLint.** For users, we want to reduce unnecessarily boilerplate and guesswork as much as possible.
+1. **Encourage plugins to use `configs`.** For plugin authors, we want to encourage the use of a consistent entrypoint, `configs`, where users can find predefined configurations. 
+
+## Detailed Design
+
+Design Summary:
+
+1. Allow arrays in config arrays (in addition to objects)
+1. Introduce an `extends` key in config objects
```

Comment:
While I fully agree with the need for a canonical way to extend configs, it's not obvious to me why the mechanism to do so should be an object key added to the config. Is that obvious? If so can it be added to this document?

The old config system worked that way, and we're all used to that, but in a static JSON config there's no other choice than an "extends" key. Hasn't that constraint been lifted? Now that eslint is all-in on the executable config system, why wouldn't just a helper function be provided instead?

Adapting the first example in this RFC, this could look like 
```js
import eslint from "eslint";
import js from "@eslint/js";
import example from "eslint-plugin-example";

export default [
  extend(
    {
      files: ["**/src/*.js"],
    },
    [js.configs.recommended, example.configs.recommended],
  ),
];
```

which actually evaluates to 

```js
import js from "@eslint/js";
import example from "eslint-plugin-example";

export default [
  [
    {
      ...js.configs.recommended,
      files: ["**/src/*.js"],
    },
    {
      ...example.configs.recommended,
      files: ["**/src/*.js"],
    },
  ],
];

```
before even being exposed from the config file, rather than assuring "if an extends key is present, it is going to be treated equivalently by eslint as if you had written out a bunch of repetitive JSON". I would think this also lets users debug any surprising behavior better.

***

Part of my motivation for this is that I _strongly_ agree with @bradzacher's every word in https://github.com/eslint/rfcs/pull/126#discussion_r1846993921. I find that the least surprising behavior regarding the `"files"` is to override the extended configs' files in an API style with an "extends" key, and I will not be budged from this perspective. That being said, there is certainly proof in this RFC that others see it differently.

A built-in helper can give the user the option, for example like so
```ts
extend(config, extensionConfigs, { files: "override" });
extend(config, extensionConfigs, { files: "intersect" });
```
And then we're only bikeshedding about what the right default is, rather than the right difficult-to-work-around behavior is.

***

Somewhat related to the previous, another concern is - as also noted several times - the "extends" key already exists in typescript-eslint's `config()` helper. This is a nontrivial user share, and it would be a real yucky situation for typescript-eslint `config()` users if `"extends"` came to have a _different_ meaning with or without wrapping the configs in `config()`. 

***

Note - one can come up with all sorts of fancy helper API designs. I only suggest `extend(config, extensionConfigs, options)` because it's the simplest. But feel free to imagine in your head whatever the most appealing style is (maybe `extend(config).with(extensionConfigs)`?). The specific signature isn't really relevant to my question.

Note also that that the `.flat(Infinity)` behavior totally makes sense to me to be handled implicitly still.

Reply:
> While I fully agree with the need for a canonical way to extend configs, it's not obvious to me why the mechanism to do so should be an object key added to the config. Is that obvious? If so can it be added to this document?

This is covered in the "Alternatives" section in the RFC. 

> Somewhat related to the previous, another concern is - as also noted several times - the "extends" key already exists in typescript-eslint's `config()` helper. 

I do realize this, and it's unfortunate, but I don't think that should prevent us from doing what makes the most sense for most users. We've heard the feedback that the way typescript-eslint recommends configuration is confusing because it looks different than the ESLint docs. Prior comments in this RFC also note that the way `extends` works in `tsconfig` doesn't always match expectations. It's unfortunate that `tsconfig` didn't at least try to mimic what `extends` did in eslintrc.

To reiterate from previous comments and the RFC itself, the overall feedback from ESLint users is that:

1) There are too many different ways to extend configs in the ecosystem
2) Beginners find using JavaScript-y ways of doing things confusing, so adding back `extends` (something familiar from eslintrc) seems to be the best approach.

Reply:
> This is covered in the "Alternatives" section in the RFC.

Ah, missed that, sorry!

> I do realize this, and it's unfortunate, but I don't think that should prevent us from doing what makes the most sense for most users. We've heard the feedback that the way typescript-eslint recommends configuration is confusing because it looks different than the ESLint docs. Prior comments in this RFC also note that the way `extends` works in `tsconfig` doesn't always match expectations. It's unfortunate that `tsconfig` didn't at least try to mimic what `extends` did in eslintrc.

Yeah I don't disagree at all with the points that the state of the typescript-eslint helper isn't ideal. All I'm suggesting is that when weighing "extends" key vs helper, the ability to avoid inconsistencies between eslint and typescript-eslint, which many users seem to assume are more or less one and the same, is _a_ benefit of the helper approach.

> To reiterate from previous comments and the RFC itself, the overall feedback from ESLint users is that:
> 
> 1. There are too many different ways to extend configs in the ecosystem

+1

> 2. Beginners find using JavaScript-y ways of doing things confusing, so adding back `extends` (something familiar from eslintrc) seems to be the best approach.

Sorry, they're asked to understand package management, module system and imports/requires, but a function call is what would put it over the edge??

Reply:
> Sorry, they're asked to understand package management, module system and imports/requires, but a function call is what would put it over the edge?

It's not necessarily the function call, but how it fits into the larger array. Trust me, I was just as surprised when we started getting feedback that people couldn't figure out how to use `.map()`, `Object.assign()`, `...`, etc., to mix and match their configs.

We need to be cognizant that ESLint is used by a huge number of developers with a wide range of experience in JavaScript.

Reply:
I hear you on this. FWIW though, I'm personally not surprised at all that users have difficulty with `.map()`, `Object.assign()`, or `...`. Those are very language-specific constructs (the last two triply so), not universal programming concepts. `.map()` is the least JS-specific, but in order to use it you need to write function expressions (arrow or not), which _are_ language-specific knowledge. Calling an existing function with concrete arguments, on the other hand, is about as universal a concept to coding as I can think of.

Reply:
But yes, to your point, how it all it all comes together is obviously an important part of the design intuitiveness as well. I think an extends helper is pretty intuitive in this context, but I'm willing to grant that that may or may not hold true to users at large

Reply:
I think overall what I'm trying to avoid here is the need for users to understand a bunch of additional tools or techniques they have to become familiar with in order to properly assemble a config array. That's why I think a solution like `defineConfig` hits a sweetspot where they can always use that, and the rest is just JavaScript arrays and objects.